### DISCUSSION TITLE
Demo question template

### DISCUSSION THREAD

**Discussion by mrdbourke** [mrdbourke/zero-to-mastery-ml#48] [General]
This template is to help with the formatting of questions to best help others who are reading it.

The four main sections are:
1. **Where you're stuck** - best to put a timestamp/video number of the course here so people know exactly where you are (e.g. ZTM video 135, timestamp 11:07)
2. **Your problem** - what is your problem?
3. **Your code** - what code did you try that isn't working?
4. **What you've tried so far** - detailing the troubleshooting steps you've taken already will help others help you.

You can copy the demo template below and fill out the sections in your question, much of the formatting has been done just to make things look nicer:

```
* Course video/timestamp: (e.g. video number 147 on ZTM, timestamp 10:35)

## My error

TODO: Add some info here about your error...

## My code

TODO: Add your code here, best to format with backticks as well, for example: 

\```python <- use triple backticks before and after your code, write "python" after the first set to make the code formatted
[your code here]
\``` <- be sure to delete the slashes, only keep the backticks

## What I've tried so far

TODO: Add some steps for what you've tried to do so far to solve your error... (this will help others know what you've done when troubleshooting) 
```

**Comment by nischal6** [comment]
* Course video/timestamp: (. video number 149 on ZTM, timestamp 10:35)

## My error

TODO: TypeError: All estimators should implement fit and transform, or can be 'drop' or 'passthrough' specifiers. '(Pipeline(steps=[('imputer', SimpleImputer(fill_value=4, strategy='constant')),
                ('onehots', OneHotEncoder(handle_unknown='ignore'))]),)' (type <class 'tuple'>) doesn't.

## My code

```python
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder

# Modelling
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split, GridSearchCV

#Setup Random seed
import numpy as np
np.random.seed(seed=42)

# Import data and drop rows with missing labels
data = pd.read_csv('car-sales-extended-missing-data.csv')
data

# Drop rows of Price column with missing data
data.dropna(subset=['Price'] , inplace=True)



# We need to change "Make, Colour" columns into numbers form then fill missing value of the all these column.
# Also we we to fill missing value of "Doors" column.

categorical_features = ['Make' , 'Colour']
categorical_transformer = Pipeline(steps=[('imputer' , SimpleImputer(strategy="constant" , fill_value="missing")),
                                          ('onehot'  , OneHotEncoder(handle_unknown="ignore"))
                                         ])
# Here, we are imputing categorical_features with Constant value of missing string helps to fill missing value.
# then we are creating OneHotEncoder changing this to numerical value.

door_features = ['Doors']
door_transformer = Pipeline(steps=[('imputer' , SimpleImputer(strategy="constant" , fill_value=4)),
                                  ]),
# We are using SimpleImputer to handle missing data on "Doors" if it has missing data then fill it with 4.

numerical_features = ['Odometer (KM)']
numeric_transformer = Pipeline(steps=[('imputer' , SimpleImputer(strategy="mean"))])
# Here it is going to fill numerical column i.e Odometer (KM) with strategy=mean. It is going to take mean value of 
# Odometer (KM) column and fill all missing rows in this respective column with mean of rest of value


# Setup preprocessing steps( fill missing value then convert to numbers)
preprocessor = ColumnTransformer(
                                  transformers = [
                                               ("cat" , categorical_transformer , categorical_features),
                                               ("door" , door_transformer , door_features ),
                                               ("num" , numeric_transformer, numerical_features)
                                                ])

# Creating a preprocessing and modelling Pipeline.
model = Pipeline(steps=[('preprocessor',preprocessor),
                        ('model' , RandomForestRegressor() )
                        ])
# Here the first step in Pipeline is to run through preprocessor then creating model of RandomForestRegressor().
# Once preprocessor is done it will build RandomForestRegressor on it.

# Split data.
X = data.drop('Price' , axis=1)
y = data['Price']

# train , test split
X_train, X_test,  y_train , y_test = train_test_split(X , y , test_size=0.2)

# Fit and score the model

model.fit(X_train , y_train)
```


\```python <- use triple backticks before and after your code, write "python" after the first set to make the code formatted
[your code here]
\``` <- be sure to delete the slashes, only keep the backticks

## What I've tried so far

TODO: Add some steps for what you've tried to do so far to solve your error... (this will help others know what you've done when troubleshooting)

**Reply by mrdbourke** [reply]
Hey @nischal6 , 

A small error in your code in this line:

```python
door_transformer = Pipeline(steps=[('imputer' , SimpleImputer(strategy="constant" , fill_value=4)),
                                  ]),
```

An extra comma in the middle/end, see the updated code here: 

```python
door_transformer = Pipeline(steps=[('imputer', SimpleImputer(strategy="constant" , fill_value=4))])
```

And the full working code here: 

```python
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder

# Modelling
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split, GridSearchCV

#Setup Random seed
import numpy as np
np.random.seed(seed=42)

# Import data and drop rows with missing labels
data = pd.read_csv('../data/car-sales-extended-missing-data.csv')
data

# Drop rows of Price column with missing data
data.dropna(subset=['Price'] , inplace=True)



# We need to change "Make, Colour" columns into numbers form then fill missing value of the all these column.
# Also we we to fill missing value of "Doors" column.

categorical_features = ['Make' , 'Colour']
categorical_transformer = Pipeline(steps=[('imputer' , SimpleImputer(strategy="constant" , fill_value="missing")),
                                          ('onehot'  , OneHotEncoder(handle_unknown="ignore"))
                                         ])
# Here, we are imputing categorical_features with Constant value of missing string helps to fill missing value.
# then we are creating OneHotEncoder changing this to numerical value.

door_features = ['Doors']
door_transformer = Pipeline(steps=[('imputer', SimpleImputer(strategy="constant" , fill_value=4))])
# We are using SimpleImputer to handle missing data on "Doors" if it has missing data then fill it with 4.

numerical_features = ['Odometer (KM)']
numeric_transformer = Pipeline(steps=[('imputer' , SimpleImputer(strategy="mean"))])
# Here it is going to fill numerical column i.e Odometer (KM) with strategy=mean. It is going to take mean value of 
# Odometer (KM) column and fill all missing rows in this respective column with mean of rest of value


# Setup preprocessing steps( fill missing value then convert to numbers)
preprocessor = ColumnTransformer(
                                  transformers = [
                                               ("cat" , categorical_transformer , categorical_features),
                                               ("door" , door_transformer , door_features ),
                                               ("num" , numeric_transformer, numerical_features)
                                                ])

# Creating a preprocessing and modelling Pipeline.
model = Pipeline(steps=[('preprocessor', preprocessor),
                        ('model', RandomForestRegressor())])

# Here the first step in Pipeline is to run through preprocessor then creating model of RandomForestRegressor().
# Once preprocessor is done it will build RandomForestRegressor on it.

# Split data.
X = data.drop('Price' , axis=1)
y = data['Price']

# train , test split
X_train, X_test,  y_train , y_test = train_test_split(X , y , test_size=0.2)

# Fit and score the model
model.fit(X_train , y_train)
```

**Comment by vfonsecal** [comment]
Handling missing data with  pandas
Following section #114 
After 

#Fill the odometer column with the mean value for the missing data
car_sales_missing["Odometer (KM)"].fillna(value=car_sales_missing["Odometer (KM)"].mean(), inplace=True)

I got the following error message:

TypeError                                 Traceback (most recent call last)
Cell In[70], line 2
      1 #Fill the odometer column with the mean value for the missing data
----> 2 car_sales_missing["Odometer (KM)"].fillna(value=car_sales_missing["Odometer (KM)"].mean(), inplace=True)

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\series.py:6225, in Series.mean(self, axis, skipna, numeric_only, **kwargs)
   6217 @doc(make_doc("mean", ndim=1))
   6218 def mean(
   6219     self,
   (...)
   6223     **kwargs,
   6224 ):
-> 6225     return NDFrame.mean(self, axis, skipna, numeric_only, **kwargs)

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\generic.py:11992, in NDFrame.mean(self, axis, skipna, numeric_only, **kwargs)
  11985 def mean(
  11986     self,
  11987     axis: Axis | None = 0,
   (...)
  11990     **kwargs,
  11991 ) -> Series | float:
> 11992     return self._stat_function(
  11993         "mean", nanops.nanmean, axis, skipna, numeric_only, **kwargs
  11994     )

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\generic.py:11949, in NDFrame._stat_function(self, name, func, axis, skipna, numeric_only, **kwargs)
  11945 nv.validate_func(name, (), kwargs)
  11947 validate_bool_kwarg(skipna, "skipna", none_allowed=False)
> 11949 return self._reduce(
  11950     func, name=name, axis=axis, skipna=skipna, numeric_only=numeric_only
  11951 )

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\series.py:6133, in Series._reduce(self, op, name, axis, skipna, numeric_only, filter_type, **kwds)
   6128     # GH#47500 - change to TypeError to match other methods
   6129     raise TypeError(
   6130         f"Series.{name} does not allow {kwd_name}={numeric_only} "
   6131         "with non-numeric dtypes."
   6132     )
-> 6133 return op(delegate, skipna=skipna, **kwds)

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\nanops.py:147, in bottleneck_switch.__call__.<locals>.f(values, axis, skipna, **kwds)
    145         result = alt(values, axis=axis, skipna=skipna, **kwds)
    146 else:
--> 147     result = alt(values, axis=axis, skipna=skipna, **kwds)
    149 return result

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\nanops.py:404, in _datetimelike_compat.<locals>.new_func(values, axis, skipna, mask, **kwargs)
    401 if datetimelike and mask is None:
    402     mask = isna(values)
--> 404 result = func(values, axis=axis, skipna=skipna, mask=mask, **kwargs)
    406 if datetimelike:
    407     result = _wrap_results(result, orig_values.dtype, fill_value=iNaT)

File ~\Desktop\sample_project_1\env\lib\site-packages\pandas\core\nanops.py:719, in nanmean(values, axis, skipna, mask)
    716     dtype_count = dtype
    718 count = _get_counts(values.shape, mask, axis, dtype=dtype_count)
--> 719 the_sum = values.sum(axis, dtype=dtype_sum)
    720 the_sum = _ensure_numeric(the_sum)
    722 if axis is not None and getattr(the_sum, "ndim", False):

File ~\Desktop\sample_project_1\env\lib\site-packages\numpy\core\_methods.py:49, in _sum(a, axis, dtype, out, keepdims, initial, where)
     47 def _sum(a, axis=None, dtype=None, out=None, keepdims=False,
     48          initial=_NoValue, where=True):
---> 49     return umr_sum(a, axis, dtype, out, keepdims, initial, where)

TypeError: unsupported operand type(s) for +: 'float' and 'method'

Tried all sources on the net as advised but cannot solve the problem.

Would you  kindly be able to clarify what is going on?

**Comment by Monishabasith** [comment]
* Course video/timestamp: video number 239 timestamp 5:51

My error
Building model with: https://tfhub.dev/google/imagenet/mobilenet_v2_130_224/classification/4
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-55-0fd4f47c95c0>](https://localhost:8080/#) in <cell line: 1>()
----> 1 model = create_model()
      2 model.summary()

2 frames
[/usr/local/lib/python3.10/dist-packages/keras/src/models/sequential.py](https://localhost:8080/#) in add(self, layer, rebuild)
     93                 layer = origin_layer
     94         if not isinstance(layer, Layer):
---> 95             raise ValueError(
     96                 "Only instances of `keras.Layer` can be "
     97                 f"added to a Sequential model. Received: {layer} "

ValueError: Only instances of `keras.Layer` can be added to a Sequential model. Received: <tensorflow_hub.keras_layer.KerasLayer object at 0x7de34da405b0> (of type <class 'tensorflow_hub.keras_layer.KerasLayer'>)

TODO: 
My code

```# Create a function which builds a Keras model
def create_model(input_shape=INPUT_SHAPE, output_shape=OUTPUT_SHAPE, model_url=MODEL_URL):
  print("Building model with:", MODEL_URL)

  # Setup the model layers
  model = tf.keras.Sequential([
    hub.KerasLayer(MODEL_URL), # Layer 1 (input layer)
    tf.keras.layers.Dense(units=OUTPUT_SHAPE,
                          activation="softmax") # Layer 2 (output layer)
  ])

  # Compile the model
  model.compile(
      loss=tf.keras.losses.CategoricalCrossentropy(),
      optimizer=tf.keras.optimizers.Adam(),
      metrics=["accuracy"]
  )

  # Build the model
  model.build(INPUT_SHAPE)

  return model```

TODO: Add your code here, best to format with backticks as well, for example: 

\```python <- use triple backticks before and after your code, write "python" after the first set to make the code formatted
[your code here]
\``` <- be sure to delete the slashes, only keep the backticks

## What I've tried so far

I tried some changes in code using stack overflow. i even tried to run your notebook of end to end dogvision. Your code also shows the same error

TODO: Add some steps for what you've tried to do so far to solve your error... (this will help others know what you've done when troubleshooting)

**Comment by BINARD300** [comment]
* Course video/timestamp: YouTube Intro to Machine Learning & Data Science in 2025 (+Pandas, NumPy, Matplotlib) 4:38:59
* https://youtu.be/r67SfaiYaDI?t=16739

## My error
<>:1: SyntaxWarning: invalid escape sequence '\$'
<>:1: SyntaxWarning: invalid escape sequence '\$'
C:\Users\w1ll0\AppData\Local\Temp\ipykernel_7444\1301473172.py:1: SyntaxWarning: invalid escape sequence '\$'
  car_sales["Price"] = car_sales["Price"].str.replace('[\$\,\.]', '').astype(int)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[236], line 1
----> 1 car_sales["Price"] = car_sales["Price"].str.replace('[\$\,\.]', '').astype(int)
      2 car_sales

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\generic.py:6665, in NDFrame.astype(self, dtype, copy, errors)
   6659     results = [
   6660         ser.astype(dtype, copy=copy, errors=errors) for _, ser in self.items()
   6661     ]
   6663 else:
   6664     # else, only a single dtype is given
-> 6665     new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
   6666     res = self._constructor_from_mgr(new_data, axes=new_data.axes)
   6667     return res.__finalize__(self, method="astype")

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\internals\managers.py:449, in BaseBlockManager.astype(self, dtype, copy, errors)
    446 elif using_copy_on_write():
    447     copy = False
--> 449 return self.apply(
    450     "astype",
    451     dtype=dtype,
    452     copy=copy,
    453     errors=errors,
    454     using_cow=using_copy_on_write(),
    455 )

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\internals\managers.py:363, in BaseBlockManager.apply(self, f, align_keys, **kwargs)
    361         applied = b.apply(f, **kwargs)
    362     else:
--> 363         applied = getattr(b, f)(**kwargs)
    364     result_blocks = extend_blocks(applied, result_blocks)
    366 out = type(self).from_blocks(result_blocks, self.axes)

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\internals\blocks.py:784, in Block.astype(self, dtype, copy, errors, using_cow, squeeze)
    781         raise ValueError("Can not squeeze with more than one column.")
    782     values = values[0, :]  # type: ignore[call-overload]
--> 784 new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
    786 new_values = maybe_coerce_values(new_values)
    788 refs = None

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\dtypes\astype.py:237, in astype_array_safe(values, dtype, copy, errors)
    234     dtype = dtype.numpy_dtype
    236 try:
--> 237     new_values = astype_array(values, dtype, copy=copy)
    238 except (ValueError, TypeError):
    239     # e.g. _astype_nansafe can fail on object-dtype of strings
    240     #  trying to convert to float
    241     if errors == "ignore":

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\dtypes\astype.py:182, in astype_array(values, dtype, copy)
    179     values = values.astype(dtype, copy=copy)
    181 else:
--> 182     values = _astype_nansafe(values, dtype, copy=copy)
    184 # in pandas we don't store numpy str dtypes, so convert to object
    185 if isinstance(dtype, np.dtype) and issubclass(values.dtype.type, str):

File ~\Desktop\BOOTCAMP1\machineLearn\sample_project1\env\Lib\site-packages\pandas\core\dtypes\astype.py:133, in _astype_nansafe(arr, dtype, copy, skipna)
    129     raise ValueError(msg)
    131 if copy or arr.dtype == object or dtype == object:
    132     # Explicit copy, or required since NumPy can't view from / to object.
--> 133     return arr.astype(dtype, copy=True)
    135 return arr.astype(dtype, copy=copy)

ValueError: invalid literal for int() with base 10: '<NA>'
TODO: Add some info here about your error...
I have tried all code in the "Introduction to pandas" handbook plus A.I.
and nothing works. Could you please help before I go mad.
Thanks in advance

## My code

\```python 
car_sales["Price"] = car_sales["Price"].str.replace('[\$\,\.]', '').astype(int)
car_sales
\``` 


TODO: Add your code here, best to format with backticks as well, for example: 

\```python <- use triple backticks before and after your code, write "python" after the first set to make the code formatted
[your code here]
\``` <- be sure to delete the slashes, only keep the backticks

## What I've tried so far
I have tried all permutations of the suggested code code (4 hrs plus) , but none have worked
TODO: Add some steps for what you've tried to do so far to solve your error... (this will help others know what you've done when troubleshooting)
